arroyo/processing/strategies/collect.py

```diff
@@ -21,7 +32,210 @@ class OffsetRange:
     timestamp: datetime
 
 
+@dataclass(frozen=True)
+class MessageBatch(Generic[TPayload]):
+    """
+    Represents a batch of messages with the range of offsets covered.
+    The range of offsets is represented as a mapping between partitions
+    and offset ranges.
+    The lowest offset in each offset range is inclusize. The highest is
+    esclusive.
+    """
+
+    messages: Sequence[Message[TPayload]]
+    offsets: Mapping[Partition, OffsetRange]
+    duration: float
+
+    def __len__(self) -> int:
+        return len(self.messages)
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__}: {len(self)} message{'s' if len(self) != 1 else ''}, open for {self.duration:0.2f} seconds>"
+
+
+class BatchBuilder(Generic[TPayload]):
```

Comment:
I don't think I like the idea of following the `submit` and `poll` approach for the BatchBuilder. 
That would make the interface look like a ProcessingStrategy while it is not. That would be confusing and it is what happens today with the collector:
https://github.com/getsentry/arroyo/blob/main/arroyo/processing/strategies/collect.py#L24

It sends me off every time when I read the code that uses it without having the class implementation in available.

I think I will try to get rid of the builder entirely

Reply:
In the end I could not remove the builder entirely. 
I changed the interface to have only two methods:
```
append
build_if_ready
```

I removed the scenario that required the force

Reply:
You do not have to use the `submit` and `poll` names if they are too close to a strategy. My reasons for suggesting something else were:

- `append` normally implies appending items to a list to me. Here you have another builder class so feels a bit different from append to me
- The main reason for suggesting the poll-like interface is that I think it is unnecessary to reinitialise the Builder class over and over for each batch. Whether or not you call it poll doesn't matter. But I think you should just have a single instance of it, as it will make BatchStep simpler.